tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1305 @@
+///     }
+///
+///     time::sleep(Duration::from_millis(50)).await;
+/// };
+///
+/// /* use the connected client */
+/// # Ok(()) }
+/// ```
+///
+/// A client will error with [std::io::ErrorKind::NotFound] for most creation
+/// oriented operations like [create] or [wait_named_pipe] unless at least once
+/// server instance is up and running at all time. This means that the typical
+/// listen loop for a server is a bit involved, because we have to ensure that
+/// we never drop a server accidentally while a client might want to connect.
+///
+/// ```no_run
+/// use std::io;
+/// use std::sync::Arc;
+/// use tokio::net::windows::NamedPipeOptions;
+/// use tokio::sync::Notify;
+///
+/// const PIPE_NAME: &str = r"\\.\pipe\named-pipe-idiomatic-server";
+///
+/// # #[tokio::main] async fn main() -> std::io::Result<()> {
+/// // The first server needs to be constructed early so that clients can
+/// // be correctly connected. Otherwise calling .wait will cause the client to
+/// // error.
+/// //
+/// // Here we also make use of `first_pipe_instance`, which will ensure that
+/// // there are no other servers up and running already.
+/// let mut server = NamedPipeOptions::new()
+///     .first_pipe_instance(true)
+///     .create(PIPE_NAME)?;
+///
+/// let shutdown = Arc::new(Notify::new());
+/// let shutdown2 = shutdown.clone();
+///
+/// // Spawn the server loop.
+/// let server = tokio::spawn(async move {
+///     loop {
+///         // Wait for a client to connect.
+///         let connected = tokio::select! {
+///             connected = server.connect() => connected,
+///             _ = shutdown2.notified() => break,
+///         };
+///
+///         // Construct the next server to be connected before sending the one
+///         // we already have of onto a task. This ensures that the server
+///         // isn't closed (after it's done in the task) before a new one is
+///         // available. Otherwise the client might error with
+///         // `io::ErrorKind::NotFound`.
+///         server = NamedPipeOptions::new()
+///             .create(PIPE_NAME)?;
+///
+///         let client = tokio::spawn(async move {
+///             /* use the connected client */
+/// #           Ok::<_, std::io::Error>(())
+///         });
+///     }
+///
+///     Ok::<_, io::Error>(())
+/// });
+/// # shutdown.notify_one();
+/// # let _ = server.await??;
+/// # Ok(()) }
+/// ```
+///
+/// [create]: NamedPipeClientOptions::create
+/// [ERROR_PIPE_BUSY]: crate::winapi::shared::winerror::ERROR_PIPE_BUSY
+/// [Windows named pipe]: https://docs.microsoft.com/en-us/windows/win32/ipc/named-pipes
+#[derive(Debug)]
+pub struct NamedPipe {
+    io: PollEvented<mio_windows::NamedPipe>,
+}
+
+impl NamedPipe {
+    /// Fallibly construct a new named pipe from the specified raw handle.
+    ///
+    /// This function will consume ownership of the handle given, passing
+    /// responsibility for closing the handle to the returned object.
```

Comment:
What happens if the function errors? It seems like the destructor of the mio type might close it?

Reply:
Yeah, it will. Only error that can be raised is the one that happens when constructing `PollEvented`, which happens after it has been infallibly converted into a `miow::Handle` which would close it on `Drop`.